oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/DefaultOAuth2AuthorizationGrant.java

```diff
@@ -100,54 +94,52 @@ class DefaultOAuth2AuthorizationGrant implements OAuth2AuthorizationGrant {
      */
     @Override
     public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
-        CompletableFuture<GrantedOAuth2AccessToken> future;
-        GrantedOAuth2AccessToken token = null;
-        for (;;) {
-            final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
-            if (!tokenFuture.isDone()) {
-                return tokenFuture;
-            }
-            if (!tokenFuture.isCompletedExceptionally()) {
-                token = tokenFuture.join();
-                if (isValidToken(token)) {
-                    return tokenFuture;
-                }
-            }
+        return tokenLoader().get();
+    }
 
-            // `tokenFuture` got completed with an invalid token; try again.
-            future = new CompletableFuture<>();
-            if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
-                break;
-            }
+    private AsyncLoader<GrantedOAuth2AccessToken> tokenLoader() {
+        if (tokenLoader != null) {
+            return tokenLoader;
         }
 
-        final CompletableFuture<GrantedOAuth2AccessToken> newTokenFuture = future;
-        if (token == null && fallbackTokenProvider != null) {
-            CompletableFuture<? extends GrantedOAuth2AccessToken> fallbackTokenFuture = null;
-            try {
-                fallbackTokenFuture = requireNonNull(
-                        fallbackTokenProvider.get(), "fallbackTokenProvider.get() returned null");
-            } catch (Exception e) {
-                logger.warn("Unexpected exception from fallbackTokenProvider.get()", e);
-            }
-            if (fallbackTokenFuture != null) {
-                fallbackTokenFuture.handle((storedToken, unused) -> {
-                    if (isValidToken(storedToken)) {
-                        newTokenFuture.complete(storedToken);
+        final Function<GrantedOAuth2AccessToken, CompletableFuture<GrantedOAuth2AccessToken>> loader =
+                token -> {
+            final CompletableFuture<GrantedOAuth2AccessToken> newTokenFuture = new CompletableFuture<>();
+
+            if (token == null && fallbackTokenProvider != null) {
+                CompletableFuture<? extends GrantedOAuth2AccessToken> fallbackTokenFuture = null;
+                try {
+                    fallbackTokenFuture = requireNonNull(
+                            fallbackTokenProvider.get(), "fallbackTokenProvider.get() returned null");
```

Comment:
Side note: It is not related to this PR but `fallbackTokenProvider` is a name that we should only use when a token acquisition fails since it is a fallback. 

I think it would be better to remove `fallback` from the API method. `tokenProvider` seems clearer.